File: src/windows/magma/app/compile/ImmutableCompileState.c
#include "./ImmutableCompileState.h"
export class ImmutableCompileState implements CompileState {
	Platform platform;
	Option<Location> findCurrentLocation;
	List<Source> sources;
	List<Import> imports;
	List<&[I8]> structureNames;
	&[I8] structures;
	&[I8] functions;
	List<Definition> definitions;
	number depth;
	constructor (Platform platform, Option<Location> findCurrentLocation, List<Source> sources, List<Import> imports, List<&[I8]> structureNames, &[I8] structures, &[I8] functions, List<Definition> definitions, number depth) {
		this.platform = platform;
		this.findCurrentLocation = findCurrentLocation;
		this.sources = sources;
		this.imports = imports;
		this.structureNames = structureNames;
		this.structures = structures;
		this.functions = functions;
		this.definitions = definitions;
		this.depth = depth;
	}
}

static CompileState createInitial() {
	return new ImmutableCompileState(Platform.Magma, new None<Location>(), Lists.empty(), Lists.empty(), Lists.empty(), "", "", Lists.empty(), 0);
}
auto temp(&[I8] anObject) {
	return Strings.equalsTo(name, anObject);
}
Bool hasLastStructureNameOf(&[I8] name) {
	return this.structureNames.findLast().filter(temp).isPresent();
}
CompileState withLocation(Location namespace) {
	return new ImmutableCompileState(this.platform, new Some<Location>(namespace), this.sources, this.imports, this.structureNames, this.structures, this.functions, this.definitions, this.depth);
}
CompileState append(&[I8] element) {
	return new ImmutableCompileState(this.platform, this.findCurrentLocation, this.sources, this.imports, this.structureNames, this.structures + element, this.functions, this.definitions, this.depth);
}
CompileState pushStructureName(&[I8] name) {
	return new ImmutableCompileState(this.platform, this.findCurrentLocation, this.sources, this.imports, this.structureNames.addLast(name), this.structures, this.functions, this.definitions, this.depth);
}
CompileState enterDepth() {
	return new ImmutableCompileState(this.platform, this.findCurrentLocation, this.sources, this.imports, this.structureNames, this.structures, this.functions, this.definitions, this.depth + 1);
}
CompileState exitDepth() {
	return new ImmutableCompileState(this.platform, this.findCurrentLocation, this.sources, this.imports, this.structureNames, this.structures, this.functions, this.definitions, this.depth - 1);
}
CompileState defineAll(List<Definition> definitions) {
	return new ImmutableCompileState(this.platform, this.findCurrentLocation, this.sources, this.imports, this.structureNames, this.structures, this.functions, this.definitions.addAll(definitions), this.depth);
}
auto temp(Definition definition) {
	return Strings.equalsTo(definition.name(), name);
}
auto temp(Definition definition1) {
	return definition1.type();
}
Option<Type> resolve(&[I8] name) {
	return this.definitions.queryReversed().filter(temp).map(temp).next();
}
CompileState clearImports() {
	return new ImmutableCompileState(this.platform, this.findCurrentLocation, this.sources, Lists.empty(), this.structureNames, this.structures, this.functions, this.definitions, this.depth);
}
CompileState clear() {
	return new ImmutableCompileState(this.platform, this.findCurrentLocation, this.sources, this.imports, this.structureNames, "", "", this.definitions, this.depth);
}
CompileState addSource(Source source) {
	return new ImmutableCompileState(this.platform, this.findCurrentLocation, this.sources.addLast(source), this.imports, this.structureNames, this.structures, this.functions, this.definitions, this.depth);
}
auto temp(&[I8] inner) {
	return Strings.equalsTo(inner, base);
}
auto temp(Source source) {
	return this.addResolvedImportWithNamespace(source.computeNamespace(), source.computeName());
}
CompileState addResolvedImportFromCache(&[I8] base) {
	if (this.structureNames.query().anyMatch(temp)) {
		return this;
	}
	return this.findSource(base).map(temp).orElse(this);
}
CompileState popStructureName() {
	return new ImmutableCompileState(this.platform, this.findCurrentLocation, this.sources, this.imports, this.structureNames.removeLast().orElse(this.structureNames), this.structures, this.functions, this.definitions, this.depth);
}
CompileState mapLocation((arg0 : Location) => Location mapper) {
	return new ImmutableCompileState(this.platform, this.findCurrentLocation.map(mapper), this.sources, this.imports, this.structureNames, this.structures, this.functions, this.definitions, this.depth);
}
CompileState withPlatform(Platform platform) {
	return new ImmutableCompileState(platform, this.findCurrentLocation, this.sources, this.imports, this.structureNames, this.structures, this.functions, this.definitions, this.depth);
}
CompileState addFunction(&[I8] function) {
	return new ImmutableCompileState(this.platform, this.findCurrentLocation, this.sources, this.imports, this.structureNames, this.structures, this.functions + function, this.definitions, this.depth);
}
&[I8] join() {
	return this.structures + this.functions;
}
auto temp(Location location) {
	return location.namespace();
}
auto temp(Import node) {
	return Strings.equalsTo(node.child(), child);
}
CompileState addResolvedImportWithNamespace(List<&[I8]> oldParent, &[I8] child) {
	List<&[I8]> namespace = this.findCurrentLocation.map(temp).orElse(Lists.empty());
	List < String > newParent = oldParent;
	if (Platform.TypeScript === this.platform) {
		if (namespace.isEmpty()) {
			newParent = newParent.addFirst(".");
		}
		number i = 0;
		number size = namespace.size();
		while (i < size) {
			newParent = newParent.addFirst("..");
			i++;
		}
	}
	if (this.imports.query().filter(temp).next().isPresent()) {
		return this;
	}
	Import importString = new Import(newParent, child);
	return new ImmutableCompileState(this.platform, this.findCurrentLocation, this.sources, this.imports.addLast(importString), this.structureNames, this.structures, this.functions, this.definitions, this.depth);
}
auto temp(Source source) {
	return Strings.equalsTo(source.computeName(), name);
}
Option<Source> findSource(&[I8] name) {
	return this.sources.query().filter(temp).next();
}
Bool isPlatform(Platform platform) {
	return platform === this.platform();
}
&[I8] createIndent() {
	number indent = this.depth();
	return "\n" + "\t".repeat(indent);
}
Option<&[I8]> findLastStructureName() {
	return this.structureNames().findLast();
}

File: src/ts/magma/app/compile/ImmutableCompileState.ts
/*[
	JVMList: jvm.api.collect.list, 
	Lists: jvm.api.collect.list, 
	Files: jvm.api.io, 
	Actual: magma.annotate, 
	Namespace: magma.annotate, 
	Collector: magma.api.collect, 
	EmptyHead: magma.api.collect.head, 
	FlatMapHead: magma.api.collect.head, 
	Head: magma.api.collect.head, 
	HeadedIter: magma.api.collect.head, 
	MapHead: magma.api.collect.head, 
	RangeHead: magma.api.collect.head, 
	SingleHead: magma.api.collect.head, 
	Iter: magma.api.collect, 
	Iters: magma.api.collect, 
	Joiner: magma.api.collect, 
	Iterable: magma.api.collect.list, 
	List: magma.api.collect.list, 
	ListCollector: magma.api.collect.list, 
	Console: magma.api.io, 
	IOError: magma.api.io, 
	Path: magma.api.io, 
	None: magma.api.option, 
	Option: magma.api.option, 
	Some: magma.api.option, 
	Err: magma.api.result, 
	Ok: magma.api.result, 
	Result: magma.api.result, 
	Characters: magma.api.text, 
	Strings: magma.api.text, 
	Tuple2: magma.api, 
	Tuple2Impl: magma.api, 
	CompileState: magma.app.compile, 
	Context: magma.app.compile, 
	ConstructionCaller: magma.app.compile.define, 
	ConstructorHeader: magma.app.compile.define, 
	Definition: magma.app.compile.define, 
	MethodHeader: magma.app.compile.define, 
	Parameter: magma.app.compile.define, 
	Dependency: magma.app.compile, 
	DivideState: magma.app.compile, 
	ImmutableCompileState: magma.app.compile, 
	ImmutableDivideState: magma.app.compile, 
	Import: magma.app.compile, 
	Registry: magma.app.compile, 
	Stack: magma.app.compile, 
	Placeholder: magma.app.compile.text, 
	Symbol: magma.app.compile.text, 
	Whitespace: magma.app.compile.text, 
	FunctionType: magma.app.compile.type, 
	PrimitiveType: magma.app.compile.type, 
	TemplateType: magma.app.compile.type, 
	Type: magma.app.compile.type, 
	VariadicType: magma.app.compile.type, 
	AccessValue: magma.app.compile.value, 
	Argument: magma.app.compile.value, 
	Caller: magma.app.compile.value, 
	Invokable: magma.app.compile.value, 
	Lambda: magma.app.compile.value, 
	Not: magma.app.compile.value, 
	Operation: magma.app.compile.value, 
	StringValue: magma.app.compile.value, 
	Value: magma.app.compile.value, 
	Compiler: magma.app, 
	PathSource: magma.app.io, 
	Source: magma.app.io, 
	Location: magma.app, 
	Main: magma.app, 
	Platform: magma.app, 
	Sources: magma.app, 
	Targets: magma.app
]*/
import { CompileState } from "../../../magma/app/compile/CompileState";
import { Context } from "../../../magma/app/compile/Context";
import { Registry } from "../../../magma/app/compile/Registry";
import { Stack } from "../../../magma/app/compile/Stack";
import { Location } from "../../../magma/app/Location";
import { Platform } from "../../../magma/app/Platform";
import { Dependency } from "../../../magma/app/compile/Dependency";
import { Lists } from "../../../jvm/api/collect/list/Lists";
import { Import } from "../../../magma/app/compile/Import";
import { Some } from "../../../magma/api/option/Some";
import { Definition } from "../../../magma/app/compile/define/Definition";
import { Iterable } from "../../../magma/api/collect/list/Iterable";
import { Source } from "../../../magma/app/io/Source";
export class ImmutableCompileState implements CompileState {
	context: Context;
	registry: Registry;
	depth: number;
	stack: Stack;
	constructor (context: Context, registry: Registry, depth: number, stack: Stack) {
		this.context/*unknown*/ = context/*Context*/;
		this.registry/*unknown*/ = registry/*Registry*/;
		this.depth/*unknown*/ = depth/*number*/;
		this.stack/*unknown*/ = stack/*Stack*/;
	}
	createIndent(): string {
		return "\n" + "\t".repeat(this.depth)/*unknown*/;
	}
	addResolvedImport(location: Location): CompileState {
		if (Platform.PlantUML === this.context().platform()/*unknown*/){
			let name = this.context().maybeLocation().map(Location.name).orElse("")/*unknown*/;
			let dependency = new Dependency(name, location.name())/*unknown*/;
			if (!this/*unknown*/.registry().dependencies().contains(dependency)/*unknown*/){
				return new ImmutableCompileState(this.context, new Registry(this.registry().imports(), this.registry().dependencies().addLast(dependency), this.registry().output()), this.depth, this.stack)/*unknown*/;
			}
		}
		let requestedNamespace = location.namespace()/*unknown*/;
		let requestedChild = location.name()/*unknown*/;
		let thisNamespace = this.context().maybeLocation().map(Location.namespace).orElse(Lists.empty())/*unknown*/;
		if (thisNamespace.isEmpty()/*unknown*/){
			requestedNamespace/*unknown*/ = requestedNamespace.addFirst(".")/*unknown*/;
		}
		let i = 0/*unknown*/;
		let size = thisNamespace.size()/*unknown*/;
		while (i < size/*unknown*/){
			requestedNamespace/*unknown*/ = requestedNamespace.addFirst("..")/*unknown*/;
			i/*unknown*/++;
		}
		if (this.registry().doesImportExistAlready(requestedChild)/*unknown*/){
			return this/*unknown*/;
		}
		let newNamespace = requestedNamespace.addLast(requestedChild)/*unknown*/;
		let importString = new Import(newNamespace, requestedChild)/*unknown*/;
		return new ImmutableCompileState(this.context, new Registry(this.registry().imports().addLast(importString), this.registry().dependencies(), this.registry().output()), this.depth, this.stack)/*unknown*/;
	}
	withLocation(location: Location): CompileState {
		return new ImmutableCompileState(new Context(this.context().platform(), new Some<Location>(location), this.context().sources()), this.registry, this.depth, this.stack)/*unknown*/;
	}
	append(element: string): CompileState {
		return new ImmutableCompileState(this.context, new Registry(this.registry().imports(), this.registry().dependencies(), this.registry().output() + element), this.depth, this.stack)/*unknown*/;
	}
	pushStructureName(name: string): CompileState {
		return new ImmutableCompileState(this.context, this.registry, this.depth, new Stack(this.stack().structureNames().addLast(name), this.stack().definitions()))/*unknown*/;
	}
	enterDepth(): CompileState {
		return new ImmutableCompileState(this.context, this.registry, this.depth + 1, this.stack)/*unknown*/;
	}
	exitDepth(): CompileState {
		return new ImmutableCompileState(this.context, this.registry, this.depth - 1, this.stack)/*unknown*/;
	}
	defineAll(definitions: Iterable<Definition>): CompileState {
		return new ImmutableCompileState(this.context, this.registry, this.depth, new Stack(this.stack().structureNames(), this.stack().definitions().addAll(definitions)))/*unknown*/;
	}
	clearImports(): CompileState {
		return new ImmutableCompileState(this.context, new Registry(Lists.empty(), this.registry().dependencies(), this.registry().output()), this.depth, this.stack)/*unknown*/;
	}
	clearOutput(): CompileState {
		return new ImmutableCompileState(this.context, new Registry(this.registry().imports(), this.registry().dependencies(), ""), this.depth, this.stack)/*unknown*/;
	}
	addSource(source: Source): CompileState {
		return new ImmutableCompileState(new Context(this.context().platform(), this.context().maybeLocation(), this.context().sources().addLast(source)), this.registry, this.depth, this.stack)/*unknown*/;
	}
	addResolvedImportFromCache(base: string): CompileState {
		if (this.stack().hasAnyStructureName(base)/*unknown*/){
			return this/*unknown*/;
		}
		return this.context().findSource(base).map((source: Source) => this.addResolvedImport(source.createLocation())/*unknown*/).orElse(this)/*unknown*/;
	}
	popStructureName(): CompileState {
		return new ImmutableCompileState(this.context, this.registry, this.depth, new Stack(this.stack().structureNames().removeLast().orElse(this.stack().structureNames()), this.stack().definitions()))/*unknown*/;
	}
	withPlatform(platform: Platform): CompileState {
		return new ImmutableCompileState(new Context(platform, this.context().maybeLocation(), this.context().sources()), this.registry, this.depth, new Stack(this.stack().structureNames().removeLast().orElse(this.stack().structureNames()), this.stack().definitions()))/*unknown*/;
	}
	context(): Context {
		return this.context/*unknown*/;
	}
	registry(): Registry {
		return this.registry/*unknown*/;
	}
	stack(): Stack {
		return this.stack/*unknown*/;
	}
}
